### PR TITLE
added csg tutorial

### PR DESCRIPTION
**Pull request by ZX-WT**
One more new PR for #1414 ;)

Please review it and I don't know how to create a list with reference to the CSG class, I think it needs some fix.

I based this tutorial on the original article and added a section to demonstrate the use of CSG. Content-wise, it may not include all the settings but should provide the basic already.

### CODE REVIEW

File: tutorials/3d/csg_tools.rst [hunk 1]
```diff
@@ -0,0 +1,191 @@
+.. _doc_csg_tools:
+
+CSG
+===
+
+CSG stands for "Constructive Solid Geometry", and is a tool to combine basic shapes or custom meshes to create more complex shapes. In the 3D modelling software, CSG is mostly known as "Boolean Operators".
```
Comment (mhilbrunner) [inline tutorials/3d/csg_tools.rst]: In 3D modelling software

File: tutorials/3d/csg_tools.rst [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+The aim of CSG in Godot is for it to be used in level prototyping. This technique allows users to create simple versions of most common shapes by combining primitives. Interior environments can be created by using inverted primitives.
+
+.. image:: img/csg.gif
+
+Introduction to CSG nodes
+-------------------
+
+Like other Godot's feature, CSG is supported in form of nodes, these are the CSG nodes
```
Comment (mhilbrunner) [inline tutorials/3d/csg_tools.rst]: Like other features of Godot, CSG is supported in the form of nodes

File: tutorials/3d/csg_tools.rst [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+- :ref:`CSGCombiner <class_CSGcombiner>'
+
+.. image:: img/csg_nodes.png
+
+.. image:: img/csg_mesh.png
+
+CSG tools features
+~~~~~~~~~~~~~~~~~~
+
+Every CSG nodes supported three kind of boolean operations
```
Comment (mhilbrunner) [inline tutorials/3d/csg_tools.rst]: Every CSG node supports three kinds

File: tutorials/3d/csg_tools.rst [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+- Union: Geometry of both primitives is merged, intersecting geometry is removed.
+- Intersection: Only intersecting geometry remains, the rest is removed.
+- Subtraction: The second shape is subtracted from the first, leaving a dent with it's shape.
```
Comment (mhilbrunner) [inline tutorials/3d/csg_tools.rst]: its

File: tutorials/3d/csg_tools.rst [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+.. image:: img/csg_poly.png
+
+.. note:: ``Path`` mode needs a ``Path`` :ref:`Path <class_Path>` node to operate. In ``Path``, draw the path and the polygon in ``CSGPolygon`` will extrude along the path.
+
+
+Custom meshes
+~~~~~~~~~~~~~
+
+Any mesh can be used for :ref:`CSGMesh <class_CSGMesh>`, the mesh can be modelled in other software and import into Godot. Multiple materials is supported. There are some restrictions for geometry:
```
Comment (mhilbrunner) [inline tutorials/3d/csg_tools.rst]: imported

File: tutorials/3d/csg_tools.rst [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+- It must be closed
+- It must not self-intersect
+- It must not contain internal faces
+- Every edge must connect to only two other faces
+
+.. image:: img/csg_custom_mesh.png
+
+CSGCombiner
+~~~~~~~~~~~
+:ref:`CSGCombiner <class_CSGCombiner>` node is an empty shape used for organization. It will only combine children nodes.
```
Comment (mhilbrunner) [inline tutorials/3d/csg_tools.rst]: *The* ... node is

File: tutorials/3d/csg_tools.rst [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+
+Process order
+~~~~~~~~~~~~~
+
+Every CSG node will first process it's children nodes and their operation: union, intersection or subtraction, in tree order and apply them to itself one after the other.
```
Comment (mhilbrunner) [inline tutorials/3d/csg_tools.rst]: its, operations

File: tutorials/3d/csg_tools.rst [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+.. note:: A note on performance : Make sure CSG geometry remains relatively simple, as complex meshes can take a while to process. If adding objects together (such as table and room objects). Please make them as separate CSG trees. Forcing too many objects in a single tree will eventually start affecting performance. Only use binary operations where you actually need them.
```
Comment (mhilbrunner) [inline tutorials/3d/csg_tools.rst]: ... room objects), please create them as separate CSG trees.

File: tutorials/3d/csg_tools.rst [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+.. image:: img/csg_shelf_subtract_menu.png
+
+You almost build a shelf. Create one more ``CSGBox`` for dividing the shelf into two levels.
+
+.. image:: img/csg_shelf.png
+
+Position your furniture in your room as you like and your scene should look this:
+
+.. image:: img/csg_room_result.png
+
```
Comment (mhilbrunner) [inline tutorials/3d/csg_tools.rst]: This ends very abruptly IMO. Add a ending sentence congratulating on finishing the tutorial?

**Comment by mhilbrunner** [comment]
This should also be added to the corresponding index (for an example, see the changes in https://github.com/godotengine/godot-docs/pull/1807)

File: tutorials/3d/csg_tools.rst [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+===
+
+CSG stands for "Constructive Solid Geometry", and is a tool to combine basic shapes or custom meshes to create more complex shapes. In the 3D modelling software, CSG is mostly known as "Boolean Operators".
+
+The aim of CSG in Godot is for it to be used in level prototyping. This technique allows users to create simple versions of most common shapes by combining primitives. Interior environments can be created by using inverted primitives.
+
+.. image:: img/csg.gif
+
+Introduction to CSG nodes
+-------------------
```
Comment (LikeLakers2) [inline tutorials/3d/csg_tools.rst]: The `-`s should extend to the same length as the header.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**